doc/reference/std/misc/atom.md

```diff
@@ -0,0 +1,339 @@
+Changes to atoms are always free of race conditions.
+
+Unlike the Clojure implementation, our current implementation does not retry,
+though we do not guarantee not to change in the future to match Clojure.
+Unlike in Clojure, it is thus acceptable for the change function to have side-effects,
+though it is not recommended, in case this behavior changes.
+
+## atom
+``` scheme
+(atom (atom (initial-value (void)) validator: (validator #f))) -> Atom
+```
+
+Creates a new atom, a box that atomically changes value.
+If no initial-value is provided, `(void)` is used (Clojure has no default initial-value).
+A `validator:` keyword argument can specify a validator
+(defaults to `#f` which designates a validator that always says "yes").
+The validator takes a new value and returns true if it is accepted,
+returns `#f` or throws if it is rejected.
+
+::: tip Examples:
+``` scheme
+> (import :std/iter :std/misc/atom)
+> (def a (atom 0))
+> (def workers (for/collect (i (in-range 100)) (spawn atom-swap! a 1+)))
+> (begin (for-each thread-join! workers) (atom-deref a))
+100
+```
+:::
+
+## atom?
+``` scheme
+(atom? a) -> boolean
+```
+
+Returns `#t` if *a* an atom, `#f` otherwise.
+
+::: tip Examples:
+``` scheme
+> (atom? (atom))
+t
+
+> (atom 42)
+#<Atom #8>
+> (atom? #8)
+#t
+```
+:::
+
+## atom-swap!
+``` scheme
+(atom-swap! atom function) -> any
+```
+
+Applies the function to the value currently in the atom, and,
+if it passes the atom's validator, sets the atom to the new value.
+Ensures that no other access is made to the atom in the middle of a swap.
+Returns the new value.
+
+::: tip Examples:
+``` scheme
+> (def a (atom 3))
+> (atom-swap! a (cut * 2 <>))
+6
+```
+:::
+
+## atom-swap-values!
+``` scheme
+(atom-swap! atom function) -> any any
+```
+
+Similar to `atom-swap!` but returns two values, the `old` and the `new` values.
+Similar to Clojure's `swap-vals!` but returns two values rather than a vector of size 2.
+
+::: tip Examples:
+``` scheme
+> (def a (atom 6))
+> (atom-swap-values! a (cut * 2 <>))
+6
+12
+```
+:::
+
+## atom-reset!
+``` scheme
+(atom-reset! atom new-value) -> new-value
+```
+
+If the new value passes the atom's validator, sets the atom to it.
+Ensures that no other access is made to the atom in the middle of a reset.
+Returns the new value.
+
+::: tip Examples:
+``` scheme
+> (def a (atom 42))
+> (atom-reset! a 23)
+23
+> (atom-deref a)
+23
+```
+:::
+
+## atom-reset-values!
+``` scheme
+(atom-reset-values! atom new-value) -> old-value new-value
+```
+
+Similar to `atom-reset!` but returns two values, the old and the new.
+Similar to Clojure `atom-reset-vals!` but returns values instead of a vector.
+
+::: tip Examples:
+``` scheme
+> (def a (atom 42))
+> (atom-reset-values! a 23)
+42
+23
+> (atom-deref a)
+23
+```
+:::
+
+## atom-compare-and-set!
+
+``` scheme
+(atom-compare-and-set! atom oldval newval) -> bool
+```
+
+Atomically sets the value of atom to newval if and only if the
```

Comment:
``` scheme
(atom-compare-and-set! atom old-val new-val) -> bool
```

Atomically sets the value of `atom` to `new-val` if and only if the current value of `atom` is `old-val`. Returns `#t` if the `set!` happened, else `#f`.